cmd/cloud-endpoints-controller/main.go

```diff
@@ -126,86 +130,54 @@ func sync(parent *CloudEndpoint, children *CloudEndpointControllerRequestChildre
-										log.Printf("[INFO][%s] Created jwtAud: %s", parent.Name, jwtAud)
-										status.JWTAudiences = append(status.JWTAudiences, jwtAud)
-									}
-								}
-							}
-							if found == false {
-								return status, &desiredChildren, fmt.Errorf("Backend not found or is not ready for service: %s, NodePort: %s", svcName, nodePort)
-							}
-						} else {
-							return status, &desiredChildren, fmt.Errorf("Service %s not type NodePort", svcName)
-						}
-					}
-				}
+		if name, key := parent.Spec.OpenAPISpecConfigMap.Name, parent.Spec.OpenAPISpecConfigMap.Key; name != "" && key != "" {
+			openAPISpecTemplate, err = getConfigMapSpecData(parent.ObjectMeta.Namespace, name, key)
+			if err != nil { //The user tried to supply a configMap spec, but it could not be loaded yet
+				log.Printf("[INFO][%s] Unable to load ConnfigMap Spec: %v", parent.Name, err)
+				return status, &desiredChildren, nil
 			} else {
-				target = parent.Spec.Target
-			}
-			openAPISpec, err = makeDefaultOpenAPISpec(status.Endpoint, target, status.JWTAudiences)
-			if err != nil {
-				return status, &desiredChildren, err
+				status.ConfigMapHash = toSha1(openAPISpecTemplate)
 			}
-			log.Printf("[INFO][%s] Target: %s", parent.Name, target)
-			status.IngressIP = target
-		} else {
-			// Use provided spec.
-			specYaml, err := yaml.Marshal(&parent.Spec.OpenAPISpec)
+		} else if parent.Spec.OpenAPISpec != nil {
+			openAPISpecTemplateBytes, err := yaml.Marshal(parent.Spec.OpenAPISpec)
 			if err != nil {
-				return status, &desiredChildren, err
+				log.Printf("[ERROR][%s] Supplied spec: %v", parent.Name, err)
+				return status, &desiredChildren, errors.Wrap(err, "Could not load user supplied API from CloudEndpoint Spec")
 			}
+			openAPISpecTemplate = string(openAPISpecTemplateBytes)
+		}
 
-			var spec map[string]interface{}
-			if err = yaml.Unmarshal([]byte(specYaml), &spec); err != nil {
+		if parent.Spec.TargetIngress.Name != "" {
+			target, status.JWTAudiences, err = getTargetIngress(parent)
+			if err != nil { //either there is a fatal error or the ingress isn't ready
+				if err != nil {
+					log.Printf("[INFO][%s] failed to get target %v", parent.Name, err)
+				} else {
+					log.Printf("[INFO][%s] target not ready.", parent.Name)
+				}
 				return status, &desiredChildren, err
 			}
-			openAPISpec = string(specYaml)
+		} else {
+			target = parent.Spec.Target
 		}
-
+		if openAPISpecTemplate == "" { // no spec from manifest of configmap
+			openAPISpecTemplate = getWildcardAPITemplate(status.JWTAudiences)
+		}
+		finalOpenAPISpec, err = executeTemplate(openAPISpecTemplate, status.Endpoint, target, status.JWTAudiences)
+		if err != nil {
+			log.Printf("[ERROR][%s] %v", parent.Name, err)
+			return status, &desiredChildren, err
+		}
+		if err := validateOpenAPISpec(finalOpenAPISpec); err != nil {
+			status.StateCurrent = StateIdle
+			return status, &desiredChildren, err
+		}
+		status.IngressIP = target
```

Comment:
`status.ingressIP` should be set earlier, after the target has be finally determined after line 167.